python/ray/_private/worker.py

```diff
@@ -867,21 +886,42 @@ def raise_errors(self, serialized_objects, object_refs):
         for e in out:
             _unhandled_error_handler(e)
 
-    def deserialize_objects(self, serialized_objects, object_refs):
+    def deserialize_objects(
+        self,
+        serialized_objects,
+        object_refs,
+        tensor_transport_hint: Optional[TensorTransportEnum] = None,
+    ):
         gpu_objects: Dict[str, List["torch.Tensor"]] = {}
         for obj_ref, (_, _, tensor_transport) in zip(object_refs, serialized_objects):
-            # If using a non-object store transport, then tensors will be sent
-            # out-of-band. Get them before deserializing the object store data.
             if (
                 tensor_transport is None
                 or tensor_transport == TensorTransportEnum.OBJECT_STORE
+            ) and (
+                obj_ref is None
+                or obj_ref.tensor_transport() == TensorTransportEnum.OBJECT_STORE.value
             ):
+                # The object is not a gpu object, so we cannot use other external transport to
+                # fetch it.
                 continue
 
+            # If the object is a gpu object, we can choose to use the object store or other external
+            # transport to fetch it. The `tensor_transport_hint` has the highest priority, so if it's
+            # `OBJECT_STORE`, we will use the object store to fetch the GPU object.
+            use_object_store = (
```

Comment:
Why use a flag for use_object_store instead of just passing the tensor transport hint value directly through? This seems non-intuitive for this case:

- user specifies ray.put(tensor_transport="nixl")
- user specifies ray.get(tensor_transport="nccl") -> transport used will still be "nixl"?